lib/web_ui/lib/src/engine/platform_dispatcher/app_lifecycle_state.dart

```diff
@@ -0,0 +1,61 @@
+// Copyright 2013 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+import 'package:meta/meta.dart';
+import 'package:ui/ui.dart' as ui;
+
+import '../../engine.dart';
+
+/// Signature of functions added as a listener to [ui.AppLifecycleState] changes
+typedef AppLifecycleStateListener = void Function(ui.AppLifecycleState state);
+
+/// Determines the [ui.AppLifecycleState].
+abstract class AppLifecycleStateSupport {
+  AppLifecycleStateSupport({required this.broadcastLifecycleState});
+
+  final AppLifecycleStateListener broadcastLifecycleState;
+
+  ui.AppLifecycleState appLifecycleState = ui.AppLifecycleState.resumed;
+
+  @visibleForTesting
+  void onAppLifecycleStateChange(ui.AppLifecycleState state) {
+    if (state != appLifecycleState) {
+      appLifecycleState = state;
+      broadcastLifecycleState(state);
+    }
+  }
+
+  void dispose();
+}
+
+/// Determines the [ui.AppLifecycleState] by listening to 'visibilitychange',
+/// 'beforeunload' events on the [DomWindow] / [DomHTMLDocument].
+class DomAppLifecycleStateSupport extends AppLifecycleStateSupport {
+  DomAppLifecycleStateSupport({required super.broadcastLifecycleState}) {
+    broadcastLifecycleState(appLifecycleState);
+    domWindow.addEventListener('beforeunload', _beforeUnloadListener);
+    domDocument.addEventListener('visibilitychange', _visibilityChangeListener);
+  }
+
+  @override
+  void dispose() {
+    domWindow.removeEventListener('beforeunload', _beforeUnloadListener);
+    domDocument.removeEventListener(
+        'visibilitychange', _visibilityChangeListener);
+  }
+
+  late final DomEventListener _beforeUnloadListener =
+      createDomEventListener((DomEvent event) {
+    onAppLifecycleStateChange(ui.AppLifecycleState.detached);
+  });
+
+  late final DomEventListener _visibilityChangeListener =
+      createDomEventListener((DomEvent event) {
+    if (domDocument.visibilityState == 'visible') {
+      onAppLifecycleStateChange(ui.AppLifecycleState.resumed);
+    } else if (domDocument.visibilityState == 'hidden') {
+      onAppLifecycleStateChange(ui.AppLifecycleState.paused);
+    }
```

Comment:
Reading through [mdn](https://developer.mozilla.org/en-US/docs/Web/API/Document/visibilityState) and Flutter [docs](https://api.flutter.dev/flutter/dart-ui/AppLifecycleState.html), it's not 100% clear to me whether this should be `paused` or `hidden`. But I think `hidden` might be better, @gspencergoog?

Reply:
Yes, I think `AppLifecycleState.hidden` makes more sense here, since that matches what the DOM is telling you. Paused is typically only entered on non-web mobile platforms.